src/main/kotlin/com/varabyte/kobweb/intellij/colors/KobwebColorProvider.kt

```diff
@@ -0,0 +1,222 @@
+                }
+
+                callee.isKobwebColorFunction("hsl(h: Float, s: Float, l: Float)") -> run {
+                    val args = element.valueArguments.evaluateArguments<Float>(3) ?: return@run
+                    return safeHslColor(args[0], args[1], args[2])
+                }
+
+                callee.isKobwebColorFunction("hsla(h: Float, s: Float, l: Float, a: Float)") -> run {
+                    val args = element.valueArguments.evaluateArguments<Float>(4) ?: return@run
+                    return safeHslColor(args[0], args[1], args[2])
+                }
+            }
+        }
+
+        else -> null
+    }
+
+    return if (currentDepth <= MAX_SEARCH_DEPTH) {
+        nextElement?.let { traceColor(it, currentDepth + 1) }
+    } else null
+}
+
+/**
+ * Pretty much emulates what CTRL-clicking on a reference would do
+ */
+private fun KtSimpleNameExpression.locateSource(): PsiElement? {
+    val resolvedMainReference = this.mainReference.resolve() ?: return null
+    val sourceDescriptor =
+        PsiNavigationSupport.getInstance().getDescriptor(resolvedMainReference) as? OpenFileDescriptor ?: return null
+    val sourcePsiFile = sourceDescriptor.file.toPsiFile(this.project) ?: return null
+    val sourceElementIdentifier = PsiUtilBase.getElementAtOffset(sourcePsiFile, sourceDescriptor.offset)
+    val sourceElement = sourceElementIdentifier.parent
+
+    return sourceElement
+}
+
+/**
+ * Uses the kotlin analysis api, as it can parse the constants much smarter.
+ * It can parse decimal, hex and binary automatically for example.
+ */
+private inline fun <reified Evaluated, reified Mapped> Collection<KtValueArgument>.evaluateArguments(
+    argCount: Int,
+    evaluatedValueMapper: (Evaluated) -> Mapped
+): Array<Mapped>? {
+    val constantExpressions = this.mapNotNull { it.getArgumentExpression() as? KtConstantExpression }
+
+    val evaluatedArguments = constantExpressions.mapNotNull {
+        analyze(it.containingKtFile) {
+            it.evaluate(KtConstantEvaluationMode.CONSTANT_LIKE_EXPRESSION_EVALUATION)?.value as? Evaluated
+        }
+    }
+
+    return if (evaluatedArguments.size != this.size || evaluatedArguments.size != argCount) null
+    else evaluatedArguments.map(evaluatedValueMapper).toTypedArray()
+}
+
+private inline fun <reified Evaluated> Collection<KtValueArgument>.evaluateArguments(argCount: Int) =
+    evaluateArguments<Evaluated, Evaluated>(argCount) { it }
+
+private fun KtNamedFunction.isKobwebColorFunction(vararg functionSignatures: String): Boolean {
+    val actualFqName = this.kotlinFqName?.asString() ?: return false
+    val actualParameters = this.valueParameterList?.text ?: return false
+    val actual = actualFqName + actualParameters
+
+    return functionSignatures.any { functionSignature ->
+        val expected = "$KOBWEB_COLOR_COMPANION_FQ_NAME.$functionSignature"
+        expected == actual
+    }
+}
+
+private fun safeRgbColor(r: Int, g: Int, b: Int) =
```

Comment:
In general, methods should be verbs, properties / variables nouns. Here, you have a method that's a noun.

Usually, methods that try do something or return null instead of throwing an exception are called "try", like you have `parse` and `tryParse`.

There are also `...orNull` methods in the standard library too.

So this should be maybe `tryCreateColor(r, g, b)`

(Of course this comment applies below across the next few lines as well)

Reply:
Agree with the naming.

> There are also ...orNull methods in the standard library too.

What exactly do you mean by that? That there is a stdlib function to do runCatching {...}.getOrNull() or that this could be done by the call site?

Reply:
For example:

https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/single-or-null.html
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/to-int-or-null.html
or one of my favorite tongue twisters: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/first-not-null-of-or-null.html

Here, though, `tryCreateColor` is probably best, since usually `...orNull` methods are queries on some instance, and here we're making a factory method which is different.

Reply:
ohh you meant as a suffix